### PR TITLE
Safely raise errors when object contains unicode

### LINKED ISSUE #20503: assert_frame_equal cannot handle differences in with unicode data
#### Code Sample, a copy-pastable example if possible

```python
# Your code here
assert_frame_equal(pd.DataFrame(['á']), pd.DataFrame(['á'])) # => True

assert_frame_equal(pd.DataFrame(['á']), pd.DataFrame(['é']))
---------------------------------------------------------------------------
UnicodeEncodeError                        Traceback (most recent call last)
<ipython-input-4-b108172c2dda> in <module>()
----> 1 assert_frame_equal(pd.DataFrame(['á']), pd.DataFrame(['é']))

.env/lib/python2.7/site-packages/pandas/util/testing.pyc in assert_frame_equal(left, right, check_dtype, check_index_type, check_column_type, check_frame_type, check_less_precise, check_names, by_blocks, check_exact, check_datetimelike_compat, check_categorical, check_like, obj)
   1311                 check_datetimelike_compat=check_datetimelike_compat,
   1312                 check_categorical=check_categorical,
-> 1313                 obj='DataFrame.iloc[:, {0}]'.format(i))
   1314
   1315

.env/lib/python2.7/site-packages/pandas/util/testing.pyc in assert_series_equal(left, right, check_dtype, check_index_type, check_series_type, check_less_precise, check_names, check_exact, check_datetimelike_compat, check_categorical, obj)
   1179                                      check_less_precise=check_less_precise,
   1180                                      check_dtype=check_dtype,
-> 1181                                      obj='{0}'.format(obj))
   1182
   1183     # metadata comparison

pandas/src/testing.pyx in pandas._testing.assert_almost_equal (pandas/src/testing.c:4156)()

pandas/src/testing.pyx in pandas._testing.assert_almost_equal (pandas/src/testing.c:3274)()

.env/lib/python2.7/site-packages/pandas/util/testing.pyc in raise_assert_detail(obj, message, left, right, diff)
   1011 {1}
   1012 [left]:  {2}
-> 1013 [right]: {3}""".format(obj, message, left, right)
   1014
   1015     if diff is not None:

UnicodeEncodeError: 'ascii' codec can't encode character u'\xe1' in position 1: ordinal not in range(128)
```
#### Problem description

The difference is caught correctly by `assert_frame_equals` but it raises an error when trying to print the differences.

#### Expected Output

```python
AssertionError: DataFrame.iloc[:, 0] are different

DataFrame.iloc[:, 0] values are different (100.0 %)
[left]:  [á]
[right]: [é]
```

#### Output of ``pd.show_versions()``

python 2.7.15
<details>

```INSTALLED VERSIONS
------------------
commit: None
python: 2.7.14.final.0
python-bits: 64
OS: Darwin
OS-release: 17.4.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: None.None

pandas: 0.22.0
pytest: None
pip: 9.0.3
setuptools: 39.0.1
Cython: None
numpy: 1.14.2
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.7.2
pytz: 2018.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
```

</details>

### PR DESCRIPTION
**Pull request by janrito**
This safely turns nd.array objects that contain unicode into a
representation that can be printed

Checklist for other PRs (remove this part if you are doing a PR for the pandas documentation sprint):

- [x] closes #20503
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by WillAyd** [comment]
This needs tests - can you add some in `pandas/tests/util/test_testing.py`?

File: doc/source/whatsnew/v0.23.0.txt [hunk 1]
```diff
@@ -1165,3 +1165,4 @@ Other
 
 - Improved error message when attempting to use a Python keyword as an identifier in a ``numexpr`` backed query (:issue:`18221`)
 - Bug in accessing a :func:`pandas.get_option`, which raised ``KeyError`` rather than ``OptionError`` when looking up a non-existant option key in some cases (:issue:`19789`)
+- Bug in :func:`raise_assert_detail` for Series and DataFrames with differing unicode data (:issue:`20503`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: this is an internal function, rather say assert_series_equal and assert_frame_equal

File: pandas/tests/util/test_testing.py [hunk 2]
```diff
@@ -276,6 +276,19 @@ def test_numpy_array_equal_message(self):
             assert_almost_equal(np.array([[1, 2], [3, 4]]),
                                 np.array([[1, 3], [3, 4]]))
 
+        expected = """numpy array are different
```
Comment (jreback) [inline pandas/tests/util/test_testing.py]: can you make a new test & indicate the gh issue with a comment

File: pandas/tests/util/test_testing.py [hunk 3]
```diff
@@ -678,6 +693,21 @@ def test_frame_equal_message(self):
                                pd.DataFrame({'A': [1, 2, 3], 'B': [4, 5, 7]}),
                                by_blocks=True)
 
+        expected = """DataFrame\\.iloc\\[:, 1\\] are different
```
Comment (jreback) [inline pandas/tests/util/test_testing.py]: same

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.23.0.txt (hunk 1 above)
Comment (janrito) [reply doc/source/whatsnew/v0.23.0.txt]: 👍 8f607c3

File: pandas/tests/util/test_testing.py (hunk 2 above)
Comment (janrito) [reply pandas/tests/util/test_testing.py]: 👍 6b087f2

File: pandas/tests/util/test_testing.py (hunk 3 above)
Comment (janrito) [reply pandas/tests/util/test_testing.py]: 👍 6b087f2

**Comment by janrito** [comment]
Thanks for the review @jreback! I've changed the message, separated the tests and added comments referencing the gh issue

**Comment by janrito** [comment]
Tests are failing on python3, not sure how to deal with the printing issue 

It seems like error messages are converted to native str type (binary in py2 and unicode in py3) so either way we encode this will add u or b prefixes to the objects

**Comment by janrito** [comment]
Ok, I'm not sure if this is the most elegant way of dealing with it, but if we encode only on python2 when the objects are base_strings, it doesn't print the text type prefixes in either python version

File: pandas/util/testing.py [hunk 4]
```diff
@@ -992,11 +992,18 @@ def raise_assert_detail(obj, message, left, right, diff=None):
         left = pprint_thing(left)
     elif is_categorical_dtype(left):
         left = repr(left)
+
```
Comment (jreback) [inline pandas/util/testing.py]: can you import PY2 and string_types up top

File: pandas/util/testing.py [hunk 5]
```diff
@@ -992,11 +992,18 @@ def raise_assert_detail(obj, message, left, right, diff=None):
...
+    if compat.PY2 and isinstance(left, compat.string_types):
+        left = left.encode('utf-8')
+
     if isinstance(right, np.ndarray):
         right = pprint_thing(right)
     elif is_categorical_dtype(right):
         right = repr(right)
 
+    if compat.PY2 and isinstance(right, compat.string_types):
```
Comment (jreback) [inline pandas/util/testing.py]: can you add a comment on these.

File: pandas/tests/util/test_testing.py [hunk 6]
```diff
@@ -499,10 +517,12 @@ def _assert_not_equal(self, a, b, **kwargs):
     def test_equal(self):
         self._assert_equal(Series(range(3)), Series(range(3)))
         self._assert_equal(Series(list('abc')), Series(list('abc')))
+        self._assert_equal(Series(list(u'áàä')), Series(list(u'áàä')))
```
Comment (jreback) [inline pandas/tests/util/test_testing.py]: can you add a test where left is unicode and right is non-unicode (but string)

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. small comments. ping on green.

File: pandas/util/testing.py (hunk 4 above)
Comment (janrito) [reply pandas/util/testing.py]: 👍575b2e8

File: pandas/util/testing.py (hunk 5 above)
Comment (janrito) [reply pandas/util/testing.py]: 👍 45d2b8e

File: pandas/tests/util/test_testing.py (hunk 6 above)
Comment (janrito) [reply pandas/tests/util/test_testing.py]: 👍 1f7e231

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @janrito nice patch!

**Pull request merged**